### ISSUE TITLE
ofxOsc receiver fails on initialization on arm7 raspberry pi 2

### ISSUE THREAD

**Issue by dstavy** [openframeworks/openFrameworks#4409]
the last message before failure  " using unimplemented  pure virtual method"
when running ofxSsc example reciever

**Comment by Maigre** [comment]
+1 
got this error too in my project

```
ofxOscReceiver oscListener;
oscListener.setup(9000);
```

throw

```
pure virtual method called
terminate called without an active exception
Aborted
```

and exit the program.

**Comment by PetrosKataras** [comment]
Not sure if/how it is related but I ve seen this error when trying to build code that uses threads on the RPi2 with clang. I didnt have this issue when building with gcc though. Here is an SO related [post](http://stackoverflow.com/questions/23583317/c-11-threads-error-pure-virtual-function-called) .

**Comment by Maigre** [comment]
it seems related to lines 83 -> 90 
of addons/ofxOsc/ofxOscReceiver.cpp

```
listen_thread = std::thread([this]{
	listen_socket->Run();
});

// detach thread so we don't have to wait on it before creating a new socket
// or on destruction, the custom deleter for the socket unique_ptr already
// does the right thing
listen_thread.detach();
```

It don't think it's the socket->Run() since executed on it's own it doesn't throw the error
(but it blocks the program of course) so it might be related to the thread call or detach.

The link provided by PetrosKataras seems to confirm that..
Another one on the same subject [https://llvm.org/bugs/show_bug.cgi?id=12730]

But sorry i am not sure how to provide further help to fix that.. i am not very familiar with those kind of compiler issues..

**Comment by arturoc** [comment]
can you try to add in your config.make, uncomment PRJECT_CFLAGS and add:

```
PROJECT_CFLAGS = -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_{1,2,4}
```

that should rebuild the project completely next time you build but if it doesn't try running `make clean` before running `make` again

**Comment by Maigre** [comment]
I did ```make clean && make -j4``` with ```PROJECT_CFLAGS = -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_{1,2,4}``` 

but i end up with the same error (at runtime i did not mention that previously)

**Comment by Maigre** [comment]
No more luck with ```-mcpu=cortex-a8``` from [https://gcc.gnu.org/bugzilla/show_bug.cgi?id=62100#c7]

**Comment by arturoc** [comment]
can you try:

```
PROJECT_CFLAGS = -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_1 -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_2 -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_4
```

**Comment by Maigre** [comment]
arf, 
i think i broke everything up
i got ```internal compiler error: Illegal instruction``` on random places now
i ll burn a new SD and try again tomorow.
keep in touch
Best,

**Comment by Maigre** [comment]
Hi,
back on this issue,
i tried 
```PROJECT_CFLAGS = -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_1 -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_2 -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_4```
and also 
```ADDON_CFLAGS = -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_1 -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_2 -D__GCC_HAVE_SYNC_COMPARE_AND_SWAP_4```
in the ofxOsc addon_config.mk

i tried those with both local and distcc compilers but ... without success.
I still end up with this 'pure virtual method called' error at runtime triggered by the osc receiver setup.

what could i try ? 
Thanks

**Comment by crx091081gb** [comment]
I have tried the same cflags and still have the same problem.

**Comment by sku11knight** [comment]
Tried the cflags and still crashing.

**Comment by mcanet** [comment]
I have tried all comments and same issues

**Comment by PetrosKataras** [comment]
I seems that this is a bug that will probably resolve with a 5.x version of gcc.. FWIW with Jessie this bug does not occur for me when building with gcc but it does occur when building with clang :) .... So if Jessie is an option for you, you could give there a try with gcc.

**Comment by Maigre** [comment]
Hi PetrosKataras, 
If i may, could you point out the way to switch from clang to gcc ?

This bug was reported from Jessie (current Raspbian) on RPi2,
so Jessie is definitely an option :)

Thanks!

**Comment by PetrosKataras** [comment]
Hey,

gcc should be the default compiler so if you haven't changed anything in your setup, OF should compile through gcc on Jessie...meaning no need to change anything...

Hope this helps!

**Comment by Maigre** [comment]
Hi, yes indeed the default compiler is GCC on Jessie, but it is gcc 4.9.
It's also the case when using the distcc mode, my Ubuntu gcc is also 4.9
Both techniques end up with this pure virtual bug..

I can't find a way right now to upgrade to gcc 5.x but i will dig into that.

**Comment by PetrosKataras** [comment]
Ok, so for you compiling directly on the RPi2 with gcc 4.9 still triggers this error.. Hmm.. Bare in mind that I havent tried to compile OF and my observations are based on a seperate project that was targeting Jessie and was using std::thread, but for me code that uses std::thread works on Jessie through gcc 4.9.

Just for the sake of it, could you try to run this minimal test case provided by the OP in [this](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=42734) bug report ?

Its just a barebone std::thread use case.

Then you can be sure that this is the exact issue and not something else.

**Comment by Maigre** [comment]
The given test passed without error..

But the difference between the test case and the actual ofxOsc thread use is that ofxOsc use <pthread.h> with "std::thread" as variable type, when the test case use <thread> with "thread" as variable type. 

I changed the ofxOsc code to use thread instead of pthread but i end up with the same error.
I guess thread and pthread are similar...

I alse tried to start a dummy function in this thread call like this:
```
// dummy function declared outside the ofxOscReceiver class
void dummy(){
     fputs("dummy()\n", stderr);
}
...
fputs("before new thread\n", stderr);
listen_thread = std::thread(dummy);   // original line 83 of ofxOscReceiver.cpp
fputs("after new thread\n", stderr);
```
but i still get this pure virtual error:
```
before new thread
after new thread
pure virtual method called
terminate called without an active exception
```

I'm getting crazy ;)

**Comment by PetrosKataras** [comment]
std::thread will use pthread on POSIX systems so its basically the same thing so I would suppose that its something on the setup/configuration/linking that is causing this since you can run a minimal example that uses std::thread but it fails when used inside OF. Unfortuanetely I am not on the RPi2 right now and this makes it a bit more difficult to figure out what could be root for this issue but I will come back to you when I get a chance to try it out.

**Comment by Maigre** [comment]
Running this example will produce the bug: https://tombarta.wordpress.com/2008/07/10/gcc-pure-virtual-method-called/

But i don't understand the link between this test case and our case:

In this test case we start a threaded function and pass an object which we destroy before the thread has a chance to use it. It end up with the thread hitting the base clase of the passed and already destroyed object.

But in our case, we start a thread without passing anything that can be destroy..
especially when doing the dummy test !

**Comment by PetrosKataras** [comment]
I think this is a different case.. Might be far fetched but at this point it doesn't hurt to try it out -- Could you comment out [this](https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworksCompiled/project/linuxarmv7l/config.linuxarmv7l.default.mk#L137) line and add a `PLATFORM_LIBRARIES += pthread` [here](https://github.com/openframeworks/openFrameworks/blob/master/libs/openFrameworksCompiled/project/linuxarmv7l/config.linuxarmv7l.default.mk#L135) ?

**Comment by Maigre** [comment]
Yes seems different..
So i replaced line 137 in openFrameworks/libs/openFrameworksCompiled/project/linuxarmv7l/config.linuxarmv7l.default.mk
from
PLATFORM_LDFLAGS += -pthread
to
PLATFORM_LIBRARIES += pthread

but... no success.. still failing at runtime with pure virtual error

**Comment by PetrosKataras** [comment]
If you dont use ofxOsc i.e just the empty example with the test code from the original bug report do you still get the error ?

**Comment by Maigre** [comment]
Ah interresting: 
the empty exemple with only the simple thread test fail too !

Here is the - almost empty - test app:
```
#include "ofApp.h"
#include <thread>

void f() {
    fputs("f()\n", stderr);
}

void ofApp::setup(){
    fputs("before new thread\n", stderr);
    thread t(f);
    fputs("after new thread\n", stderr);
    t.join();
    fputs("after join\n", stderr);
}
...
```

It fails with the default config
It fails with the modification proposed in config.linuxarmv7l.default.mk
It also fails with the different PROJECT_CFLAGS proposed in this issue thread

**Comment by PetrosKataras** [comment]
Ok, so at least we know that its something OF specific :) . As I said I would personally expect to be some kind of configuration/linking mismatch/conflict thats causing this.. I ll try to have a look later when I have the RPi2 at hand.

**Comment by Maigre** [comment]
Yes sorry i am kind of a lost with this config/linking/compiler questions..
I could setup a RPi2 with an internet ssh access for you if you want 
(and if you have time to work on it of course)
let me know by PM :)

**Comment by Maigre** [comment]
OK there is a new development:
I was preparing a clean and up-to-date RPi2 to allow you to remote access it (details sent by email)
and now the emptyApp with a simple thread call is working without error !

But the openFrameworks/examples/addons/oscReceiveExample/
still produce this pure virtual error.. 

I will retry differents solution now that the basic threading seems to work..

**Comment by Maigre** [comment]
Wow that's really weird,
now the simple thread test is failing again... 
It was previously running fine on my up to date RPi, and did success on my other RPi after firmware update so i thought it was related to the firmware version, but now the test is failing again on both devices..

There is some zombies or destructor/constructor timing issue here..

**Comment by Maigre** [comment]
OK!
I think i found the trick, thanks to this Chess Engine github ;)
https://github.com/mAarnos/Hakkapeliitta/issues/1

The needed flag for ofxOsc (and in any situation using threads with OF)  is -march=armv6
so you should add 
```
ADDON_CFLAGS = -march=armv6
```
into **openFrameworks/addons/ofxOsc/addon_config.mk**

It looks a bit like a dirty trick since it seems to force compiler to armv6 on a armv7 plateform, but now ofxOscReceiver is working properly ! I don't know how it is supposed to be added permanently to OF..

Thanks for your help !

**Comment by jvcleave** [comment]
Bizarre. I tested on RPI/armv6 and it is ok there at least.

On the RPI2 I was able to run with with @Maigre workaround but in a different place

`PROJECT_CFLAGS = -march=armv6`

in 
`openFrameworks/examples/addons/oscReceiveExample/config.make`

I guess it's an ok workaround since armv6 binaries can run on armv7 fine. My cross-compiler gave me some warnings

`warning: switch -mcpu=cortex-a7 conflicts with -march=armv6 switch`

but it worked just fine

**Comment by arturoc** [comment]
we should probably detect raspian + gcc < 5 and set this flag in the general makefile config, otherwise threads won't work. it might make some things slower but i think it's worth changing it since the OS binaries are compiled for armv6 anyway i think

**Comment by jvcleave** [comment]
@arturoc that looks right

` readelf -A /usr/lib/arm-linux-gnueabihf/libc.a`

produces
```
File: /usr/lib/arm-linux-gnueabihf/libc.a(dl-tlsdesc.o)
Attribute Section: aeabi
File Attributes
  Tag_CPU_name: "6"
  Tag_CPU_arch: v6
  Tag_ARM_ISA_use: Yes
  Tag_THUMB_ISA_use: Thumb-1
  Tag_FP_arch: VFPv2
  Tag_ABI_align_needed: 8-byte
  Tag_ABI_align_preserved: 8-byte, except leaf SP
  Tag_ABI_HardFP_use: SP and DP
  Tag_ABI_VFP_args: VFP registers
```

It seems you would have to be using "Official" Debian armhf in order to get Armv7 support [more here](http://sjoerd.luon.net/posts/2015/02/debian-jessie-on-rpi2/)

One weird note while troubleshooting this - if you run

`strace ./oscReceiveExample`

it doesn't crash

**Comment by arturoc** [comment]
ok, perhaps we can detect the architecture for which the system was build by using readelf on a very common binary? i guess this is an exception to other systems. also i wonder if we should recommend using archlinux on the PI instead of raspbian. i've used it several times and it works really well and it's compiled for armv7 (i think even with neon ) which could make a difference for things like opencv

**Comment by jvcleave** [comment]
Hard to say without testing but if there are no benefits (along with stability issues) in using arm7 with Raspbian maybe we just point RPi users to the OF armv6 version.

I do think NEON is available on the RPI2 however. For stuff like OpenCv would probably have to compile it yourself. I've compiled libav with NEON support for the RPI2 but haven't really tested it much.

**Comment by arturoc** [comment]
mmh, yeah that make sense using the arm7v version with raspbian won't make much difference probably and doesn't seem a good idea to mix architectures.

what i meant with neon is that i think packages in archlinux already come compiled with neon optimizations (we use opencv from the system we don't compile it ourselves anymore) so it can made a difference between using arm6 fpu optimized library and armv7 with neon specially for things like opencv but also gstreamer...
